xop.go

```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
 	}
 }
 
-func getFieldByName(element reflect.Value, name string) (reflect.Value, error) {
-	for i := 0; i < element.NumField(); i++ {
-		rTypeField := element.Type().Field(i)
-		rValueField := element.Field(i)
+func getFieldFromPath(val reflect.Value, path []string) (reflect.Value, error) {
+	val = unwrapValue(val)
 
-		if rTypeField.Name == name {
-			return rValueField, nil
-		}
-	}
-	return reflect.Value{}, errFieldNotFound
-}
-
-// This assumes, if it encounters an array field, that it will be filling in the first field and continuing
-// TODO: support arbitrary array size.
-func getFieldByXMLTagPrefix(element reflect.Value, prefix string) (reflect.Value, error) {
-	if element.Kind() == reflect.Interface || element.Kind() == reflect.Ptr {
-		element = element.Elem()
-	}
-	if element.Kind() == reflect.Array || element.Kind() == reflect.Slice {
-		return getFieldByXMLTagPrefix(element.Index(0), prefix)
-	}
-	if element.Kind() != reflect.Struct {
+	// val must be a struct and path must have length > 0
+	if val.Type().Kind() != reflect.Struct || len(path) == 0 {
 		return reflect.Value{}, errFieldNotFound
 	}
 
-	for i := 0; i < element.NumField(); i++ {
-		rTypeField := element.Type().Field(i)
-		rValueField := element.Field(i)
+	// search the struct fields with the path
+	for i := 0; i < val.NumField(); i++ {
+		typeField := val.Type().Field(i)
+		valueField := val.Field(i)
+		tag := typeField.Tag.Get("xml")
+
+		// skip the XMLName field
+		if typeField.Name == xmlName {
+			continue
+		}
 
-		if tagValue, tagOk := rTypeField.Tag.Lookup("xml"); tagOk {
-			if !strings.HasPrefix(tagValue, prefix) {
-				continue
+		// skip omitted fields
+		if tag == "-" {
+			continue
+		}
+
+		// unwrap the value
+		valueField = unwrapValue(valueField)
+
+		// check if the value was unwrapped completely
+		if valueField.Type().Kind() == reflect.Slice || valueField.Type().Kind() == reflect.Slice || valueField.Type().Kind() == reflect.Ptr {
+			// if valueField is in path
+			if getNameFromTag(tag) == path[0] {
+				// if valueField is the desired field, return
+				if len(path) == 1 {
+					return valueField, nil
+				}
+
+				// valueField is empty, there is nothing more to search
```

Comment:
I'm not sure I understand this case. Why is valueField empty if len(path) != 1? What if it's >1?

Reply:
`unwrapValue` returns an array or slice only if it has no elements, and therefore cannot be unwrapped further. It similarly returns a pointer only if it is a nil pointer. At this point we've unwrapped valueField and found that the result is either an array, a slice or a pointer.

If valueField is the last value in path then it is the field we've been searching for, so we return it. If len(path) > 1, then valueField is not the last value in path, so we return `errFieldNotFound`.